### DISCUSSION TITLE
Document build result usage in functions

### DISCUSSION THREAD

**Discussion by theKashey** [vercel/vercel#4010] [General] [answered]
There is a handy model for SPAs, when there is no "real" SSR, but some information like `featureFlags`, `pageTitle`/`description` and visa versa are inserted "on render".

Usually, it is working like:
- read `public/index.html`
- read some stuff from other sources
- make a text replace
- send HTML. Done

### Problems
- function and application are being built in a parallel, so build results do not exist for the functions.
- path to `index` during the dev is `../public/index.html` and `../index.html` at the prod environment (unreachable anyway)
- no postprocess could copy build results anywhere to `includeFiles`, as long as build process is parallelized.

### Solution
I haven't found anything better than read `index.html` by actually reading it:
- rename `index.html` to something more unique, like `main.html`
- in `index.js` function `await (await fetch(`${req.headers['x-forwarded-proto']}://${http://req.headers.host}/main.html`)).text();`, ie read from "yourself"

That is working, and probably working well, especially with some request caching applied. But is it a supposed way of doing things?

### More questions
- SSR code splitting requires assess to `stats.json` or `manifest.json`, which are again - results of the build. How to access them? How `next` is assesing them?
- Is it a good idea to use "public"/"https" host to read own resources? It is possible to provide a "local" host name to make `own-read` more efficient?

**Comment by coetry** [comment, marked as answer]
Our Next.js builder is open source and can be found [here](https://github.com/zeit/now/tree/master/packages/now-next). 

We have some [custom build utilities](https://github.com/zeit/now/tree/master/packages/now-build-utils) that allow us to customize how we create serverless functions including which files need to be present when creating them.

Feel free to explore creating your own [Custom Runtime](https://github.com/zeit/now/blob/master/DEVELOPING_A_RUNTIME.md) that meets your individual needs!

**Comment by joa-queen** [comment]
@theKashey were you able to make this work? Can you provide an example?

I'm struggling with the same challenge. Documentation is not clear enough.

**Reply by theKashey** [reply]
All I needed was my own `index.html` and here is the code to get it

```js
const getIndex = async (proto: string, host: string) => (
   (await fetch(`${proto}://${host}/main.html`)).text()
)

const indexHTML = await getIndex(String(req.headers['x-forwarded-proto'] || 'http'), String(req.headers.host));
```

I am using this pattern more or less often, utilising vercel edge cache

**Reply by joa-queen** [reply]
I was trying to get the built index locally, to avoid the latency. I'll use your solution otherwise.

Thanks for the response!

**Reply by theKashey** [reply]
It's actually doing that "locally". The "host" would be some internal AWS host

**Reply by joa-queen** [reply]
Excellent! Thanks a lot!
